### PR TITLE
[quant][graphmode] Quantizing traced modules

### PR DESCRIPTION
**Pull request by jerryzh168**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #39919 [wip][quant] Support general op modules with inplace options
* #39925 [quant][graphmode] Support quantizing `repeat`
* #39924 [quant][graphmode] Support squeeze/unsqueeze
* #39923 [quant][graphmode] Run RemoveRedundantDequantize in the end
* #39915 [quant][graphmode] Pass debug option into insert_quant_dequant pass
* #39911 [quant][graphmode] Use quantized::batch_norm in graph mode
* #39910 [quant] add quantized::batch_norm
* #39895 [quant][graphmode] Support prim::TupleUnpack and prim::TupleConstruct
* #39892 [quant][graphmode][refactor] isScalar check
* #39867 [jit][subgraph_rewriter] Support list of filters
* **#39826 [quant][graphmode] Quantizing traced modules**

Summary:
Expanding operator test coverage to traced modules

Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D21991266](https://our.internmc.facebook.com/intern/diff/D21991266)

### CODE REVIEW

File: aten/src/ATen/native/quantized/cpu/qmul.cpp [hunk 1]
```diff
@@ -145,6 +145,27 @@ class QMulScalarOut final {
   }
 };
 
+template <bool ReLUFused = false>
+class QMulScalarTensor final {
+ public:
+  static Tensor run(Tensor qa, Tensor b) {
+    TORCH_CHECK(qa.qscheme() == kPerTensorAffine ||
+              qa.qscheme() == kPerTensorSymmetric,
+              "Only per tensor quantization is suuported in Mul.");
```
Comment (vkuzo) [inline aten/src/ATen/native/quantized/cpu/qmul.cpp:154]: nit: `supported`?

File: torch/testing/_internal/common_quantization.py [hunk 2]
```diff
@@ -130,16 +130,27 @@ def wrapper(*args, **kwargs):
             fn(*args, **kwargs)
     return wrapper
 
+def get_script_module(model, tracing, data):
```
Comment (vkuzo) [inline torch/testing/_internal/common_quantization.py:133]: is this used?

File: aten/src/ATen/native/quantized/cpu/qadd.cpp [hunk 3]
```diff
@@ -249,6 +249,16 @@ Tensor qadd_scalar_out(Tensor qa, Scalar b, Tensor out) {
   return _add_scalar_out<ReLUFused>(out, qa, b);
 }
 
+template <bool ReLUFused = false>
+Tensor qadd_scalar_tensor(Tensor qa, Tensor b) {
```
Comment (vkuzo) [inline aten/src/ATen/native/quantized/cpu/qadd.cpp:253]: wondering if there is context for this...b is a tensor which represents a scalar?  Is this related to how tracing works?  Maybe we can add a comment?

File: aten/src/ATen/native/quantized/cpu/qadd.cpp [hunk 4]
```diff
@@ -249,6 +249,16 @@ Tensor qadd_scalar_out(Tensor qa, Scalar b, Tensor out) {
...
+  return qadd_scalar(qa, b.item());
+}
+
+template <bool ReLUFused = false>
+Tensor qadd_scalar_tensor_out(Tensor qa, Tensor b, Tensor out) {
```
Comment (vkuzo) [inline aten/src/ATen/native/quantized/cpu/qadd.cpp:258]: same, if this is necessary can we add a comment on the reason?

File: torch/csrc/jit/passes/quantization/insert_observers.cpp [hunk 5]
```diff
@@ -636,33 +636,52 @@ graph(%self, %input, %batchnorm):
     %second_output = aten::relu_(%first_output)
     return (%second_output) )", {is_batchnorm2d_module});
 
-
   const PatternInfo mul_nn_relu = PatternInfo::parse_from_str(R"(
-graph(%self, %a, %b):
+graph(%self, %a, %b, %relu):
```
Comment (vkuzo) [inline torch/csrc/jit/passes/quantization/insert_observers.cpp:640]: more for my own understanding, this is not a logic change and just moving `%relu` out, right?

**Review by vkuzo — APPROVED** [review]
accepting to unblock.  Would be great if we added some comments inline on the why behind these changes, i.e. what specifically is different between the graphs generated by scripting vs tracing.

File: aten/src/ATen/native/quantized/cpu/qadd.cpp (hunk 3 above)
Comment (jerryzh168) [reply aten/src/ATen/native/quantized/cpu/qadd.cpp:253]: yeah, will add a comment. tracing will trace a Scalar as tensor

Comment (vkuzo) [reply aten/src/ATen/native/quantized/cpu/qadd.cpp:253]: makes sense, thx!

File: torch/csrc/jit/passes/quantization/insert_observers.cpp (hunk 5 above)
Comment (jerryzh168) [reply torch/csrc/jit/passes/quantization/insert_observers.cpp:640]: this is a logic change, we are improving the support here to be able to match relu modules used multiple times and the matching in `is_relu_module` is also more accurate than `match::module["ReLU"]`

File: torch/testing/_internal/common_quantization.py (hunk 2 above)
Comment (jerryzh168) [reply torch/testing/_internal/common_quantization.py:133]: this is used in test, you probably missed the changes in `test_quantize_script.py`, it's too large so the change is not displayed by default in github..

File: aten/src/ATen/native/quantized/cpu/qadd.cpp (hunk 3 above)
Comment (raghuramank100) [reply aten/src/ATen/native/quantized/cpu/qadd.cpp:256]: Wondering if the comment is accurate: add and add_scalar cannot be merged as add_scalar does not require a scale and zero-point.

Comment (jerryzh168) [reply aten/src/ATen/native/quantized/cpu/qadd.cpp:256]: it's a little bit confusing, I meant when we add boradcasting op, we can merge add_scalar with that op, a separate task is to merge all variations of quantized::add into `quantized::add`, since we have overloading.

File: test/quantization/test_quantize_script.py [hunk 6]
```diff
@@ -1233,80 +1232,42 @@ def _test_op_impl(self, module, data, quantized_op, tracing=False, debug=False,
...
+    @skipIfNoFBGEMM
+    def test_quantized_conv_relu(self):
         """tests for conv1d_relu/conv2d_relu/conv3d_relu"""
         conv_module = {1 : torch.nn.Conv1d, 2 : torch.nn.Conv2d, 3 : torch.nn.Conv3d}
 
         class ConvNdRelu(torch.nn.Module):
             def __init__(self, dim, inplace):
                 super(ConvNdRelu, self).__init__()
-                self.conv = conv_module[dim](1, 4, 2, 3).float()
+                self.conv = conv_module[dim](3, 3, 3).float()
```
Comment (raghuramank100) [inline test/quantization/test_quantize_script.py:1270]: Why is the dimension changed here?

**Review by raghuramank100 — APPROVED** [review]

File: test/quantization/test_quantize_script.py (hunk 6 above)
Comment (jerryzh168) [reply test/quantization/test_quantize_script.py:1270]: you mean the argument? they are changed because I changed the input dimensions
